runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/LyftProcessJobBundleFactory.java

```diff
@@ -0,0 +1,169 @@
+    private final IdGenerator idGenerator;
+    private final ControlClientPool.Source clientSource;
+
+    private LyftPythonEnvironmentFactory(
+        JobInfo jobInfo,
+        GrpcFnServer<FnApiControlClientPoolService> controlServiceServer,
+        GrpcFnServer<GrpcLoggingService> loggingServiceServer,
+        GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer,
+        GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer,
+        IdGenerator idGenerator,
+        ControlClientPool.Source clientSource) {
+      this.jobInfo = jobInfo;
+      this.processManager = ProcessManager.create();
+      this.controlServiceServer = controlServiceServer;
+      this.loggingServiceServer = loggingServiceServer;
+      this.retrievalServiceServer = retrievalServiceServer;
+      this.provisioningServiceServer = provisioningServiceServer;
+      this.idGenerator = idGenerator;
+      this.clientSource = clientSource;
+    }
+
+    /** Creates a new, active {@link RemoteEnvironment} backed by a forked process. */
+    @Override
+    public RemoteEnvironment createEnvironment(RunnerApi.Environment environment) throws Exception {
+      String workerId = idGenerator.getId();
+
+      String pipelineOptionsJson = JsonFormat.printer().print(jobInfo.pipelineOptions());
+      HashMap<String, String> env = new HashMap<>();
+      env.put("WORKER_ID", workerId);
+      env.put("PIPELINE_OPTIONS", pipelineOptionsJson);
+      env.put(
+          "LOGGING_API_SERVICE_DESCRIPTOR",
+          loggingServiceServer.getApiServiceDescriptor().toString());
+      env.put(
+          "CONTROL_API_SERVICE_DESCRIPTOR",
+          controlServiceServer.getApiServiceDescriptor().toString());
+      env.put("SEMI_PERSISTENT_DIRECTORY", "/tmp");
+
+      String sdkHarnessEntrypoint = "apache_beam.runners.worker.sdk_worker_main";
+      String executable = "bash";
+      //List<String> args = ImmutableList.of("-c", String.format("'env; python -m %s'", sdkHarnessEntrypoint));
+      List<String> args =
+          ImmutableList.of("-c", String.format("env; python -m %s", sdkHarnessEntrypoint));
+
+      LOG.info("Creating Process with ID {}", workerId);
+      // Wrap the blocking call to clientSource.get in case an exception is thrown.
+      InstructionRequestHandler instructionHandler = null;
+      try {
+        processManager.startProcess(workerId, executable, args, env);
+        // Wait on a client from the gRPC server.
+        while (instructionHandler == null) {
+          try {
+            instructionHandler = clientSource.take(workerId, Duration.ofMinutes(2));
+          } catch (TimeoutException timeoutEx) {
+            LOG.info(
+                "Still waiting for startup of environment '{}' for worker id {}",
+                executable,
+                workerId);
+          } catch (InterruptedException interruptEx) {
+            Thread.currentThread().interrupt();
```

Comment:
Any idea why would it get interrupted? is it user initiated or the environment initiated interrupt?